proof/refine/ARM/Ipc_R.thy

```diff
@@ -2304,20 +2296,478 @@ crunches doIPCTransfer, possibleSwitchTo
+  unfolding replyPush_def updateReply_def bind_assoc
+  apply (wpsimp wp: schedContextDonate_valid_objs' updateReply_valid_objs'
+                    hoare_vcg_if_lift2 threadGet_wp hoare_vcg_imp_lift')
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  apply (safe; (fastforce simp: obj_at'_def projectKOs valid_tcb_state'_def)?)
+  by (insert reply_ko_at_valid_objs_valid_reply';
+      fastforce simp: valid_reply'_def obj_at'_def projectKOs valid_bound_obj'_def)+
+
+lemma replyPush_valid_queues[wp]:
+  "\<lbrace>valid_queues and valid_objs'\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_. valid_queues\<rbrace>"
+  supply if_split [split del]
+  unfolding replyPush_def updateReply_def bind_assoc
+  apply (rule hoare_seq_ext[OF _ stateAssert_inv])
+  apply simp
+  apply (rule hoare_seq_ext[OF _ tg_sp'])
+  apply (rule hoare_seq_ext[OF _ tg_sp'])
+  apply (rule hoare_seq_ext[OF _ get_reply_sp'])
+  apply (wpsimp wp: schedContextDonate_valid_queues
+                    threadGet_wp hoare_vcg_if_lift2 hoare_vcg_imp_lift')
+  apply (clarsimp simp: ko_at_obj_at'[where P=\<top>])
+  apply (clarsimp simp: valid_reply'_def dest!: reply_ko_at_valid_objs_valid_reply')
+  done
+
+crunches reply_unlink_tcb
+  for sc_replies_sc_at[wp]: "\<lambda>s. Q (sc_replies_sc_at P scp s)"
+  (wp: crunch_wps)
+
+lemma BlockedOnReply_sym_ref1:
+  "\<lbrakk>st_tcb_at ((=) (Structures_A.thread_state.BlockedOnReply rp)) thread s; sym_refs (state_refs_of s)\<rbrakk>
+  \<Longrightarrow> reply_tcb_reply_at ((=) (Some thread)) rp s"
+  apply (clarsimp simp: pred_tcb_at_def reply_tcb_reply_at_def)
+  apply (drule (1) sym_refs_obj_atD[where p=thread])
+  apply (clarsimp simp: state_refs_of_def obj_at_def tcb_st_refs_of_def
+                 split: Structures_A.thread_state.splits)
+  apply (rename_tac ko; case_tac ko; clarsimp simp: get_refs_def2)
+  done
+
+lemma BlockedOnReply_sym_ref2:
+  "\<lbrakk>sym_refs (state_refs_of s);
+    valid_replies s;
+    reply_tcb_reply_at ((=) (Some scp)) rp s;
+    (rp, scp') \<in> replies_with_sc s\<rbrakk> \<Longrightarrow>
+    st_tcb_at (\<lambda>st. st = Structures_A.thread_state.BlockedOnReply rp) scp s"
+  apply (frule (1) valid_repliesD1_simp, clarsimp simp: replies_blocked_def)
+  apply (subgoal_tac "t = scp")
+   apply (clarsimp simp: pred_tcb_at_def obj_at_def)
+  apply (subgoal_tac "reply_tcb_reply_at ((=) (Some t)) rp s")
+   apply (clarsimp simp: reply_tcb_reply_at_def obj_at_def)
+   apply (case_tac "reply_tcb r"; clarsimp)
+  apply (rule BlockedOnReply_sym_ref1; clarsimp simp: pred_tcb_at_def obj_at_def)
+  done
+
+lemma no_replySC_valid_replies'_sc_asrt:
+  "replySCs_of s r = None \<Longrightarrow> valid_replies'_sc_asrt r s"
+  unfolding valid_replies'_sc_asrt_def
+  by (simp)
+
+crunches doIPCTransfer
+  for replySCs_of[wp]: "\<lambda>s. P (replySCs_of s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+(* FIXME RT: replace get_ep_valid_ep with this better version *)
```

Comment:
It probably lives in AInvs, so it would also involve changes to AInvs. By default I try to restrict these PRs to lib+Refine only, but don't need to. Do you think it's worth updating AInvs while I'm here?

Reply:
Maybe just to `Move_R` then? If nothing else, that makes it accessible to the RISCV64 proofs as well.